File: GEMV/Python/GeneratedCode.py

# numpy

# Prompt: "Create 3 code suggestions using the following parameters: Python numpy general matrix-vector multiply def."

# SUGGESTION 1 #

import numpy as np

def matrix_vector_multiply(matrix, vector):
    """
    Returns the dot product of a matrix and a vector.
    
    Parameters:
        matrix (numpy array): The matrix for the multiplication.
        vector (numpy array): The vector for the multiplication.
    
    Returns:
        numpy array: The result of the matrix-vector multiplication.
    """
    return np.dot(matrix, vector)


# SUGGESTION 2 #

import numpy as np

def dynamic_matrix_vector_multiply(matrix, vector, num_iterations=1000):
    """
    Returns the result of repeatedly multiplying a matrix and a vector.
    
    Parameters:
        matrix (numpy array): The matrix for the multiplication.
        vector (numpy array): The vector for the multiplication.
        num_iterations (int, optional): The number of times to repeat the multiplication. Defaults to 1000.
    
    Returns:
        numpy array: The result of the repeated matrix-vector multiplication.
    """
    result = np.zeros((num_iterations,))
    for i in range(num_iterations):
        result[i] = np.dot(matrix, vector)
    return result


# SUGGESTION 3 #

import numpy as np
from numba import jit

@jit
def matrix_vector_multiply(matrix, vector):
    """
    A faster version of the matrix-vector multiplication using NumBA JIT compiler.
    
    Parameters:
        matrix (numpy array): The matrix for the multiplication.
        vector (numpy array): The vector for the multiplication.
    
    Returns:
        numpy array: The result of the matrix-vector multiplication.
    """
    return np.dot(matrix, vector)

# ------------------------------------------------------------------------------------------------------------------------------------------------------

# cuPy

# Prompt: "Create 3 code suggestions using the following parameters: Python cuPy general matrix-vector multiply def."

# SUGGESTION 1 #

import cupy

def cupy_matrix_vector_multiply(A, b):
    """
    Perform matrix-vector multiplication using CuPy.
    
    Parameters:
        A (cupy.ndarray): The matrix for the multiplication.
        b (cupy.ndarray): The vector for the multiplication.
    
    Returns:
        cupy.ndarray: The result of the matrix-vector multiplication.
    """
    return cupy.matmul(A, b)


# SUGGESTION 2 #

import cupy

def batched_cuppy_matrix_vector_multiply(A, b, batch_size=16):
    """
    Perform batched matrix-vector multiplication using CuPy.
    
    Parameters:
        A (cupy.ndarray): The matrix for the multiplication.
        b (cupy.ndarray): The vector for the multiplication.
        batch_size (int, optional): The size of the batch. Defaults to 16.
    
    Returns:
        cupy.ndarray: The result of the batched matrix-vector multiplication.
    """
    # Split the inputs into batches
    A_batches = cupy.split(A, batch_size, axis=0)
    b_batches = cupy.split(b, batch_size, axis=0)
    
    # Perform batched matrix-vector multiplication
    results = []
    for batch_a, batch_b in zip(A_batches, b_batches):
        results.append(cupy.matmul(batch_a, batch_b))
    
    # Concatenate the results
    return cupy.concatenate(results)


# SUGGESTION 3 #

import cupy

def strided_cuppy_matrix_vector_multiply(A, b, stride=1):
    """
    Perform strided matrix-vector multiplication using CuPy.
    
    Parameters:
        A (cupy.ndarray): The matrix for the multiplication.
        b (cupy.ndarray): The vector for the multiplication.
        stride (int, optional): The stride of the matrix. Defaults to 1.
    
    Returns:
        cupy.ndarray: The result of the strided matrix-vector multiplication.
    """
    # Get the shape of the matrix and vector
    m, n = A.shape
    k = b.shape[1]
    
    # Create a strided version of the matrix
    A_strided = cupy.lib.stride_tricks.as_strided(A, shape=(m, n), strides=(stride, 1))
    
    # Perform strided matrix-vector multiplication
    return cupy.matmul(A_strided, b)

# ------------------------------------------------------------------------------------------------------------------------------------------------------

# pyCUDA

# Prompt: "Create 3 code suggestions using the following parameters: Python pyCUDA general matrix-vector multiply def."

# SUGGESTION 1 #

import pycuda.driver as cudadrv
from pycuda.compiler import SourceModule

# Define the kernel
mod = SourceModule("""
__global__ void matrix_vector_mult(float *A, float *B, float *C, int N) {
    int i = blockIdx.x * blockDim.x + threadIdx.x;
    if (i < N) {
        C[i] = dot(A, B, i);
    }
}

__shared__ float dot(float *A, float *B, int i) {
    float sum = 0;
    for (int j = 0; j < N; ++j) {
        sum += A[i * N + j] * B[j];
    }
    return sum;
}
 """)

# Load the kernel
k = mod.get_function("matrix_vector_mult")

# Initialize data
N = 8
A = np.random.rand(N, N).astype(np.float32)
B = np.random.rand(N, 1).astype(np.float32)
C = np.zeros((N, 1), dtype=np.float32)
dA = cudadrv.to_device(A)
dB = cudadrv.to_device(B)
dC = cudadrv.to_device(C)

# Launch kernel
block_size = 16
grid_size = (N + block_size - 1) // block_size
k(dA, dB, dC, N, block_size, grid_size)

# Synchronize
cudadrv.synchronize()

# Copy result back to host
C = dC.copy_to_host()


# SUGGESTION 2 #

import pycuda.driver as cudadrv
from pycuda.compiler import SourceModule

# Define the kernel
mod = SourceModule("""
__global__ void matrix_vector_mult(float *A, float *B, float *C, int N) {
    int i = blockIdx.x * blockDim.x + threadIdx.x;
    if (i < N) {
        C[i] = dot(A, B, i);
    }
}

__shared__ float dot(float *A, float *B, int i) {
    float sum = 0;
    for (int j = 0; j < N; ++j) {
        sum += A[i * N + j] * B[j];
    }
    return sum;
}
 """)

# Load the kernel
k = mod.get_function("matrix_vector_mult")

# Initialize data
N = 8
A = np.random.rand(N, N).astype(np.float32)
B = np.random.rand(N, 1).astype(np.float32)
C = np.zeros((N, 1), dtype=np.float32)
dA = cudadrv.to_device(A)
dB = cudadrv.to_device(B)
dC = cudadrv.to_device(C)

# Launch kernel
block_size = 16
grid_size = (N + block_size - 1) // block_size
k(dA, dB, dC, N, block_size, grid_size)

# Synchronize
cudadrv.synchronize()

# Copy result back to host
C = dC.copy_to_host()


# SUGGESTION 3 #

import pycuda.driver as cudadrv
from pycuda.compiler import SourceModule

# Define the kernel
mod = SourceModule("""
__global__ void matrix_vector_mult(float *A, float *B, float *C, int N) {
    __shared__ float smem[N];
    int i = blockIdx.x * blockDim.x + threadIdx.x;
    if (i < N) {
        smem[i] = dot(A, B, i);
        __syncthreads();
        C[i] = smem[i];
    }
}

__shared__ float dot(float *A, float *B, int i) {
    float sum = 0;
    for (int j = 0; j < N; ++j) {
        sum += A[i * N + j] * B[j];
    }
    return sum;
}
 """)

# Load the kernel
k = mod.get_function("matrix_vector_mult")

# Initialize data
N = 8
A = np.random.rand(N, N).astype(np.float32)
B = np.random.rand(N, 1).astype(np.float32)
C = np.zeros((N, 1), dtype=np.float32)
dA = cudadrv.to_device(A)
dB = cudadrv.to_device(B)
dC = cudadrv.to_device(C)

# Launch kernel
block_size = 16
grid_size = (N + block_size - 1) // block_size
k(dA, dB, dC, N, block_size, grid_size)

# Synchronize
cudadrv.synchronize()

# Copy result back to host
C = dC.copy_to_host()

# ------------------------------------------------------------------------------------------------------------------------------------------------------

# Numba

# Prompt: "Create 3 code suggestions using the following parameters: Python Numba general matrix-vector multiply def."

# SUGGESTION 1 #

import numpy as np
from numba import jit

@jit(nopython=True)
def matvec_mul(A, b):
    """
    Matrix-vector multiplication.
    
    Parameters
    ----------
    A : ndarray
        Input matrix.
    b : ndarray
        Input vector.
        
    Returns
    -------
    out : ndarray
        Output vector.
    """
    m, n = A.shape
    out = np.zeros((m,))
    for i in range(m):
        for j in range(n):
            out[i] += A[i, j] * b[j]
    return out


# SUGGESTION 2 #

import numpy as np
from numba import jit

@jit(nopython=True)
def matvec_mul(As, bs):
    """
    Element-wise matrix-vector multiplication.
    
    Parameters
    ----------
    As : ndarray
        Array of input matrices.
    bs : ndarray
        Array of input vectors.
        
    Returns
    -------
    out : ndarray
        Array of output vectors.
    """
    m, n = As.shape
    out = np.zeros((len(As), n))
    for i in range(len(As)):
        for j in range(n):
            out[i, j] = np.sum(As[i] * bs[j])
    return out


# SUGGESTION 3 #

import numpy as np
from numba import jit, prange, parallel

@jit(nopython=True)
@parallel
def matvec_mul(As, bs):
    """
    Parallel element-wise matrix-vector multiplication.
    
    Parameters
    ----------
    As : ndarray
        Array of input matrices.
    bs : ndarray
        Array of input vectors.
        
    Returns
    -------
    out : ndarray
        Array of output vectors.
    """
    m, n = As.shape
    out = np.zeros((m, bs.size))

    # Split the computation into smaller chunks
    chunksize = 100
    nchunks = (bs.size + chunksize - 1) // chunksize

    @jit(nopython=True)
    def chunk_matvec_mul(A, b, out):
        """
        Computes the matrix-vector product for a single chunk.
        
        Parameters
        ----------
        A : ndarray
            Input matrix.
        b : ndarray
            Input vector.
        out : ndarray
            Output vector.
        """
        for i in prange(chunksize):
            for j in range(n):
                out[i, j] = np.dot(A[i], b[j])

    # Perform the computation in parallel
    with parallel():
        for chunk in range(nchunks):
            chunk_matvec_mul(As[:, chunk*chunksize:(chunk+1)*chunksize], bs[chunk*chunksize:(chunk+1)*chunksize], out[:, chunk*chunksize:(chunk+1)*chunksize])

    return out

# ------------------------------------------------------------------------------------------------------------------------------------------------------

File: GEMV/Julia/GeneratedCode.jl
# ===================================================================================================================
#               Threads Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia Threads general matrix-vector multiply."

# SUGGESTION 1 # ----------------------------------

using Threads

# Create a matrix and vector
A = Matrix(rand(10, 10))
b = Vector(rand(10))

# Perform matrix-vector multiplication using threads
@threads for i in 1:10
    A[i, :] * b
end


# SUGGESTION 2 # ----------------------------------

using Parallel

# Create a matrix and vector
A = Matrix(rand(10, 10))
b = Vector(rand(10))

# Perform matrix-vector multiplication using parallel blocks
@parallel for i in 1:10
    A[i, :] * b
end


# SUGGESTION 3 # ----------------------------------

using Distributed

# Create a matrix and vector
A = Matrix(rand(10, 10))
b = Vector(rand(10))

# Split the matrix and vector into smaller chunks
A_chunks = split(A, (5, 5))
b_chunks = split(b, (5,))

# Perform matrix-vector multiplication in parallel on multiple processes
@distributed for chunk in A_chunks
    @distributed for j in 1:length(chunk)
        chunk[j, :] * b_chunks[j]
    end
end



# ===================================================================================================================
#               CUDA Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia CUDA general matrix-vector multiply."

# SUGGESTION 1 # ----------------------------------

using CUDA

# Define the matrix and vector
A = CuMatrix(rand(Float32, 10, 10))
b = CuVector(rand(Float32, 10))

# Perform matrix-vector multiplication
C = A * b


# SUGGESTION 2 # ----------------------------------

using CUDA

# Define the matrix and vector with strides
A = CuMatrix(rand(Float32, 10, 10), stride=2)
b = CuVector(rand(Float32, 10), stride=2)

# Perform matrix-vector multiplication with strides
C = A[:,:] * b[::2]


# SUGGESTION 3 # ----------------------------------

using CUDA

# Define the matrix and vector
A = CuMatrix(rand(Float32, 10, 10))
b = CuVector(rand(Float32, 10))

# Allocate shared memory for the matrix-vector product
smem = CuSharedMemory( Float32, 10 )

# Perform matrix-vector multiplication using shared memory
for i in 1:10
    # Load elements from global memory to shared memory
    smem[i] = A[i,:] * b[i]
end

# Store the result in a new CuMatrix object
C = smem[:]



# ===================================================================================================================
#               AMDGPU Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia AMDGPU general matrix-vector multiply."

# SUGGESTION 1 # ----------------------------------

using AMDGPU

# Define the matrix and vector
A = AMDGPU.zeros((10, 10), Float32)
b = AMDGPU.zeros((10,), Float32)

# Perform matrix-vector multiplication
C = A * b


# SUGGESTION 2 # ----------------------------------

using AMDGPU

# Define the matrix and vector with strides
A = AMDGPU.zeros((10, 10), Float32, stride=2)
b = AMDGPU.zeros((10,), Float32, stride=2)

# Perform matrix-vector multiplication with strides
C = A[:,:] * b[::2]


# SUGGESTION 3 # ----------------------------------

using AMDGPU

# Define the matrix and vector
A = AMDGPU.zeros((10, 10), Float32)
b = AMDGPU.zeros((10,), Float32)

# Allocate shared memory for the matrix-vector product
smem = AMDGPU.shared_memory( Float32, 10 )

# Perform matrix-vector multiplication using shared memory
for i in 1:10
    # Load elements from global memory to shared memory
    smem[i] = A[i,:] * b[i]
end

# Store the result in a new AMDGPU array
C = smem[:]



# ===================================================================================================================
#               KernelAbstractions Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia KernelAbstractions general matrix-vector multiply."

# SUGGESTION 1 # ----------------------------------

using KernelAbstractions

# Define the matrix and vector
A = KernelAbstractions.Matrix(undef, (10, 10), Float32)
b = KernelAbstractions.Vector(undef, (10,), Float32)

# Perform matrix-vector multiplication
C = A * b


# SUGGESTION 2 # ----------------------------------

using KernelAbstractions

# Define the matrix and vector with strides
A = KernelAbstractions.Matrix(undef, (10, 10), Float32, stride=2)
b = KernelAbstractions.Vector(undef, (10,), Float32, stride=2)

# Perform matrix-vector multiplication with strides
C = A[:,:] * b[::2]


# SUGGESTION 3 # ----------------------------------

using KernelAbstractions

# Define the matrix and vector
A = KernelAbstractions.Matrix(undef, (10, 10), Float32)
b = KernelAbstractions.Vector(undef, (10,), Float32)

# Allocate shared memory for the matrix-vector product
smem = KernelAbstractions.shared_memory( Float32, 10 )

# Perform matrix-vector multiplication using shared memory
for i in 1:10
    # Load elements from global memory to shared memory
    smem[i] = A[i,:] * b[i]
end

# Store the result in a new kernel abstraction object
C = smem[:]



# ---------------------------------------------------------------------------------------------------------------------------------------------------------


